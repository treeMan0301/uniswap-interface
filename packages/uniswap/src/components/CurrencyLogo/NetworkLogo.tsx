import React from 'react'
import { Flex, FlexProps, Image, useSporeColors } from 'ui/src'
import { iconSizes } from 'ui/src/theme'
import { UNIVERSE_CHAIN_INFO } from 'uniswap/src/constants/chains'
import { ChainId } from 'uniswap/src/types/chains'

type NetworkLogoProps = FlexProps & {
  chainId: ChainId
  size?: number
  shape?: 'circle' | 'square'
}

export const SQUARE_BORDER_RADIUS = 6

export function TransactionSummaryNetworkLogo({
  chainId,
  size = iconSizes.icon20,
}: Pick<NetworkLogoProps, 'chainId' | 'size'>): JSX.Element {
  return (
    <Flex
      borderColor="$surface1"
      borderRadius={7} // when inner icon borderRadius = 6 and size = 20
      borderWidth={2}>
      <NetworkLogo chainId={chainId} shape="square" size={size} />
    </Flex>
  )
}

function _NetworkLogo({
  chainId,
  shape,
  size = iconSizes.icon20,
}: NetworkLogoProps): JSX.Element | null {
  const logo = UNIVERSE_CHAIN_INFO[chainId]?.logo
  const colors = useSporeColors()
  const borderRadius = shape === 'circle' ? size / 2 : SQUARE_BORDER_RADIUS
  return logo ? (
    <Flex
      style={{ borderColor: colors.surface1.get(), borderRadius, overflow: 'hidden' }}
      testID="network-logo">
      <Image resizeMode="contain" source={logo} style={{ width: size, height: size }} />
    </Flex>
  ) : null
}

export const NetworkLogo = React.memo(_NetworkLogo)
