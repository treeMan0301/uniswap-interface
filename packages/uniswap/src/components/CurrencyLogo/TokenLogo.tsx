import { memo } from 'react'
import { Flex, Text, UniversalImage, useColorSchemeFromSeed, useSporeColors } from 'ui/src'
import { iconSizes, validColor } from 'ui/src/theme'
import { STATUS_RATIO } from 'uniswap/src/components/CurrencyLogo/CurrencyLogo'
import { NetworkLogo } from 'uniswap/src/components/CurrencyLogo/NetworkLogo'
import { UniverseChainId } from 'uniswap/src/types/chains'

interface TokenLogoProps {
  url?: string | null
  symbol?: string
  name?: string | null
  chainId?: UniverseChainId
  size?: number
  hideNetworkLogo?: boolean
  networkLogoBorderWidth?: number
}

export const TokenLogo = memo(function _TokenLogo({
  url,
  symbol,
  name,
  chainId,
  size = iconSizes.icon40,
  hideNetworkLogo,
  networkLogoBorderWidth = 1.5,
}: TokenLogoProps): JSX.Element {
  const colors = useSporeColors()
  const { foreground, background } = useColorSchemeFromSeed(name ?? symbol ?? '')

  const showNetworkLogo = !hideNetworkLogo && chainId && chainId !== UniverseChainId.Mainnet
  const networkLogoSize = Math.round(size * STATUS_RATIO)

  const fallback = (
    <Flex
      alignItems="center"
      borderRadius="$roundedFull"
      height={size}
      justifyContent="center"
      px="$spacing8"
      style={{ backgroundColor: background }}
      width={size}
    >
      <Text
        adjustsFontSizeToFit
        $platform-web={{
          // adjustFontSizeToFit is a react-native-only prop
          fontSize: 10,
          whiteSpace: 'nowrap',
          overflow: 'hidden',
          textOverflow: 'clip',
        }}
        allowFontScaling={false}
        color={validColor(foreground)}
        fontFamily="$button"
        fontSize={17}
        fontWeight="500"
        lineHeight={14}
        minimumFontScale={0.5}
        numberOfLines={1}
      >
        {symbol?.slice(0, 3)}
      </Text>
    </Flex>
  )

  const tokenImage = (
    <UniversalImage
      fallback={fallback}
      size={{ height: size, width: size }}
      style={{
        image: {
          backgroundColor: colors.white.val,
          borderRadius: size / 2,
        },
      }}
      testID="token-image"
      uri={url ?? undefined}
    />
  )

  return (
    <Flex alignItems="center" height={size} justifyContent="center" testID="token-logo" width={size}>
      {tokenImage}
      {showNetworkLogo && (
        <Flex bottom={-2} position="absolute" right={-3}>
          <NetworkLogo borderWidth={networkLogoBorderWidth} chainId={chainId} size={networkLogoSize} />
        </Flex>
      )}
    </Flex>
  )
})
