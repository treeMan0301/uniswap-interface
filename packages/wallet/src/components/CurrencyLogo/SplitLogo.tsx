import { Flex } from 'ui/src'
import { CurrencyLogo, STATUS_RATIO } from 'uniswap/src/components/CurrencyLogo/CurrencyLogo'
import { TransactionSummaryNetworkLogo } from 'uniswap/src/components/CurrencyLogo/NetworkLogo'
import { CurrencyInfo } from 'uniswap/src/features/dataApi/types'
import { ChainId } from 'uniswap/src/types/chains'

interface Props {
  inputCurrencyInfo: Maybe<CurrencyInfo>
  outputCurrencyInfo: Maybe<CurrencyInfo>
  size: number
  chainId: ChainId | null
}

/*
 * Logo, where left 50% of width is taken from one icon (its left 50%)
 * and right side is taken from another icon (its right 50%)
 */
export function SplitLogo({
  size,
  inputCurrencyInfo,
  outputCurrencyInfo,
  chainId,
}: Props): JSX.Element {
  const iconSize = size / 2

  const icon =
    chainId && chainId !== ChainId.Mainnet ? (
      <TransactionSummaryNetworkLogo chainId={chainId} size={size * STATUS_RATIO} />
    ) : undefined

  return (
    <Flex height={size} width={size}>
      <Flex
        left={0}
        overflow="hidden"
        position="absolute"
        testID="input-currency-logo-container"
        top={0}
        width={iconSize - 1 /* -1 to allow for space between the icons */}>
        <CurrencyLogo hideNetworkLogo currencyInfo={inputCurrencyInfo} size={size} />
      </Flex>
      <Flex
        flexDirection="row-reverse"
        overflow="hidden"
        position="absolute"
        right={0}
        testID="output-currency-logo-container"
        top={0}
        width={iconSize - 1 /* -1 to allow for space between the icons */}>
        <CurrencyLogo hideNetworkLogo currencyInfo={outputCurrencyInfo} size={size} />
      </Flex>
      {icon && (
        <Flex bottom={-4} position="absolute" right={-4}>
          {icon}
        </Flex>
      )}
    </Flex>
  )
}
