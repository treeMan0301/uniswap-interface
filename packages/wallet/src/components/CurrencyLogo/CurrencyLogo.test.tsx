import { CurrencyLogo } from 'uniswap/src/components/CurrencyLogo/CurrencyLogo'
import {
  ARBITRUM_DAI_CURRENCY_INFO,
  UNI_CURRENCY_INFO,
  arbitrumDaiCurrencyInfo,
} from 'wallet/src/test/fixtures'
import { render } from 'wallet/src/test/test-utils'

describe(CurrencyLogo, () => {
  it('renders without error', () => {
    const tree = render(<CurrencyLogo currencyInfo={UNI_CURRENCY_INFO} />)

    expect(tree).toMatchSnapshot()
  })

  it('renders a currency logo with network logo', () => {
    const tree = render(<CurrencyLogo currencyInfo={ARBITRUM_DAI_CURRENCY_INFO} />)

    expect(tree).toMatchSnapshot()
  })

  describe('network logo', () => {
    it('is rendered by default', () => {
      const { queryByTestId } = render(<CurrencyLogo currencyInfo={arbitrumDaiCurrencyInfo()} />)

      const networkLogo = queryByTestId('network-logo')

      expect(networkLogo).toBeTruthy()
    })

    it('is rendered if hideNetworkLogo is false', () => {
      const { queryByTestId } = render(
        <CurrencyLogo currencyInfo={arbitrumDaiCurrencyInfo()} hideNetworkLogo={false} />
      )

      const networkLogo = queryByTestId('network-logo')

      expect(networkLogo).toBeTruthy()
    })

    it('is not rendered if hideNetworkLogo is true', () => {
      const { queryByTestId } = render(
        <CurrencyLogo hideNetworkLogo currencyInfo={arbitrumDaiCurrencyInfo()} />
      )

      const networkLogo = queryByTestId('network-logo')

      expect(networkLogo).toBeFalsy()
    })
  })
})
