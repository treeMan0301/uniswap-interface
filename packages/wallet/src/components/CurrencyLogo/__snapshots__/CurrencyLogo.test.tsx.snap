// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyLogo renders a currency logo with network logo 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "column",
      "height": 40,
      "justifyContent": "center",
      "width": 40,
    }
  }
  testID="token-logo"
>
  <Image
    resizeMode="contain"
    source={
      {
        "uri": "https://raw.githubusercontent.com/Uniswap/assets/master/blockchains/ethereum/assets/0x6B175474E89094C44Da98b954EedeAC495271d0F/logo.png",
      }
    }
    style={
      [
        {
          "backgroundColor": "#2222220D",
          "borderColor": "#2222220D",
          "borderRadius": 20,
          "borderWidth": 0.5,
          "height": 40,
          "width": 40,
        },
      ]
    }
    testID="token-image"
  />
  <View
    style={
      {
        "borderBottomColor": "#FFFFFF",
        "borderBottomLeftRadius": 8,
        "borderBottomRightRadius": 8,
        "borderBottomWidth": 2,
        "borderLeftColor": "#FFFFFF",
        "borderLeftWidth": 2,
        "borderRightColor": "#FFFFFF",
        "borderRightWidth": 2,
        "borderStyle": "solid",
        "borderTopColor": "#FFFFFF",
        "borderTopLeftRadius": 8,
        "borderTopRightRadius": 8,
        "borderTopWidth": 2,
        "bottom": -2,
        "flexDirection": "column",
        "position": "absolute",
        "right": -3,
      }
    }
  >
    <View
      style={
        {
          "borderColor": "#FFFFFF",
          "borderRadius": 6,
          "flexDirection": "column",
          "overflow": "hidden",
        }
      }
      testID="network-logo"
    >
      <Image
        resizeMode="contain"
        source={
          {
            "process": [Function],
          }
        }
        style={
          {
            "height": 16,
            "width": 16,
          }
        }
      />
    </View>
  </View>
</View>
`;

exports[`CurrencyLogo renders without error 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "column",
      "height": 40,
      "justifyContent": "center",
      "width": 40,
    }
  }
  testID="token-logo"
>
  <Image
    resizeMode="contain"
    source={
      {
        "uri": "https://raw.githubusercontent.com/Uniswap/assets/master/blockchains/ethereum/assets/0x1f9840a85d5aF5bf1D1762F925BDADdC4201F984/logo.png",
      }
    }
    style={
      [
        {
          "backgroundColor": "#2222220D",
          "borderColor": "#2222220D",
          "borderRadius": 20,
          "borderWidth": 0.5,
          "height": 40,
          "width": 40,
        },
      ]
    }
    testID="token-image"
  />
</View>
`;
